### PR TITLE
Profit sharing for captured vessels

### PR DESCRIPTION
**Pull request by ravenshining**
### Context

Money is easy to make, and the player doesn't have a lot in the way of
overheads. This makes it very simple for a player to make a lot of money
in a short space of time, which allows them to quickly amass a powerful
fleet and do away with the game's sense of challenge.

We've been adding mechanics to help mitigate this issue. For the last
few versions, commodity trading has required the player to pay profit
shares to the various crew across the fleet. This helps slows down
economic advancement once the player has amassed some power, but leaves
it relatively unhindered in the early game.

We need to extend profit sharing to other kinds of income, such as
capturing and plundering enemy ships. Currently, capturing and
plundering are by far the most powerful way to make money. Capturing is
far more profitable than plundering, especially in the early game.

### Changes

This commit makes capturing and plundering add an outstanding debt to
the player's bank account. The amount of the debt is equal to the value
of the captured or plundered assets, multiplied by the fleet's current
profit share ratio.

### Considerations

As of this commit, the interest rate and term of the profit sharing debt
are hard-coded. In a future commit, we intend to make those values based
on some settings in the data files.

In the future, we want to introduce system that tracks the morale of
each ship in the fleet and creates incentives for the player to keep
their crew alive and happy. As part of that, we intend to punish the
loss of life that often takes place when inherent in capturing ships by
having it create morale problems.

We plan for the profit sharing debts to also affect morale. How it does
that exactly will be up to the data files. At this stage, we think that
having outstanding profit sharing debt would cause morale to tick down
each day, but paying the profit shares up front would trigger a morale
boost. Being unable to pay the daily rates on the profit sharing debt
would of course be a morale disaster.

### CODE REVIEW

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]: Rather than piggybacking on fines, I think this would be worth adding as its own entry.

It's useful for the player to be able to see that they have outstanding profit shares, and it'll become even more useful in the future when these things affect morale.

**Comment by lukearndt** [comment]
How would you feel about adding just the profit shares, instead of bundling them together with death benefits?

File: source/Account.cpp
Comment (ravenshining) [reply source/Account.cpp]: I agree completely with this and will look into it.  This piggybacking is just how ES did this way back when DBs were a thing.

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]: This is a good start, but the calculation that we'll actually need is more complex than this.

Have a look at `Crew.h` and `Crew.cpp`; that's where we currently keep this kind of logic. You may be able to reuse some of the existing functionality.

Comment (ravenshining) [reply source/BoardingPanel.cpp]: Yep, that's on the TODO!

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]: I think a higher interest rate would be better, to encourage the player not to let them hang around for too long.

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]: I expect we're paying profit shares to the whole fleet, so perhaps we could just say "your crew"?

File: source/Account.cpp
Comment (lukearndt) [reply source/Account.cpp]: Actually, the interest rate and term need to be defined by a data files. Maybe the game start settings?

Comment (ravenshining) [reply source/Account.cpp]: I agree, but Kiko wanted *no* interest, so I thought this was good compromise?

**Comment by kikotheexile** [comment]
i am more concerned about how and that it functions, opposed to the terminology or verbiage used. although both can always be changed at a later date, there's no reason to hold up *testing* of a function over a few choice words that can be ironed out during testing.

File: source/BoardingPanel.cpp
Comment (ravenshining) [reply source/BoardingPanel.cpp]: Only the crewmembers of the flagship risk themselves in the actual boarding, and that's who I *think* I'm calculating against, but that is a good question.

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]: Normally when we want to share profits, we pass the gross income into `Crew::ShareProfit` and let the Crew system handle it for us. However, in this case we're going to need a two-stage approach because we don't want the profit to actually get shared just yet. Instead, we want to calculate how much profit we _need_ to share.

One of the reasons for this is that we want to be able to call a convenient function and have the system handle everything for us. In the future, this will be super useful for making sure that all of the correct things happen; for example, morale increases in response to the profit being shared.

Currently, `Crew::ShareProfit` just returns the amount of profit that we need to share. That's because the morale system doesn't exist in yet in the master branch.

I propose that we add a function called `Crew::CalculateProfitShare` with the same arguments as `Crew::ShareProfit` and cut+paste the contents of `Crew::ShareProfit` into it. Then we make `Crew::ShareProfit` call `Crew::CalculateProfitShare` under the hood for now.

Then this code would be:

```suggestion
		profitShare += Crew::CalculateProfitShare(
			player.Ships(),
			player.Flagship(),
			plunder[selected].UnitValue() * currentShares * Depreciation::Full()
		);
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
```
With the other suggestions, this line of code becomes unnecessary because we won't be calculating profit shares in this file.

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
				// By taking the ship, you have earned profit. You must share it.
				profitShares += Crew::CalculateProfitShare(
					player.Ships(),
					player.Flagship(),
					victim->Cost() * startShares * Depreciation::Full()
				);
```

File: source/Account.h
Comment (lukearndt) [inline source/Account.h]:
```suggestion
	void AddProfitShares(int64_t amount);
```

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
void Account::AddProfitShares(int64_t amount)
{
	mortgages.emplace_back(amount, 1000, 90);
}
```

Just for consistency with the other variable names in this file.

File: source/Account.cpp
Comment (lukearndt) [reply source/Account.cpp]: Let's let the data files decide. 😄

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
	int64_t profitSharesPaid = 0;
```

This may seem nitpicky, but I've found that naming things exactly what they are makes the code way easier to understand down the track.

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
				profitSharesPaid += payment;
```

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
	if(profitSharesPaid)
		typesPaid["profit shares"] = profitSharesPaid;
```

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
				<< ((mortgagesPaid || finesPaid || maintenancePaid || profitSharesPaid) ? " and " : ".");
```

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp]:
```suggestion
				<< ((mortgagesPaid || finesPaid || profitSharesPaid) ? " and " : ".");
		if(sharesPaid)
			out << creditString(profitSharesPaid) << " in outstanding profit shares"
				<< ((finesPaid || mortgagesPaid) ? " and " : ".");
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
			Messages::Add(
				"You must pay " +
				Format::Number(profitShares) +
				" credits to your fleet as profit shares."
			);
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp:239]: Nice. I agree - let's let the `Accounts` system handle it from here.

File: source/BoardingPanel.h
Comment (lukearndt) [inline source/BoardingPanel.h]:
```suggestion
```
The other suggestions make this unnecessary.

File: source/Account.cpp
Comment (lukearndt) [reply source/Account.cpp]: I added the word `outstanding` to help communicate that the crew would really prefer if we paid them their profit shares right away.

File: source/Account.cpp
Comment (ravenshining) [reply source/Account.cpp]: I did wonder at that when I saw it in MZ's code.  I didn't touch it since that's how he did it, but if you say it's safe yeah your way makes more sense to me.

File: source/BoardingPanel.cpp
Comment (lukearndt) [reply source/BoardingPanel.cpp]: The profit sharing should absolutely take the whole fleet into account. The fleet enabled the player to board that ship in the first place by protecting the flagship, risking their lives in the process.

All escorts deserve a share of the profits.

Comment (lukearndt) [reply source/BoardingPanel.cpp]: It's worth noting that one of the primary design goals of the profit sharing system is to make it less attractive for the player to have an overly large and powerful fleet. Smaller player fleets mean more risky combat encounters, which in turn makes for more exciting gameplay.

File: source/Account.cpp
Comment (ravenshining) [reply source/Account.cpp]: Sure, though I don't know how to do that yet.  Also as you can see in Mortgage.cpp below, the game determines what to call a debt based on its interest rate.  Not an ideal way of doing it, I know, but if you let the data change this then how that works will have to be changed as well.

File: source/BoardingPanel.cpp
Comment (ravenshining) [reply source/BoardingPanel.cpp]: Hmm.  You're right that all escorts contributed to making the boarding action possible.... but at the same time, the crewmembers of the flagship undertake, ignoring the meta of the player's ship is always going to survive, a much greater risk.  Not only do they have to sit in a capture barge that might be more vulnerable to attack while the victim is disabled, but the chances of someone dying during boarding are much greater.  It seems to me they ought to be paid more, corresponding to their greater risk.

Well, unless you're the sort of player who goes around with Sparrow fleets, then the fleet is probably taking a similar risk.

File: source/BoardingPanel.cpp
Comment (ravenshining) [reply source/BoardingPanel.cpp]: Sure, that language suits either model, but I liked having a specific mention of the boarding action for clarity.

**Comment by ravenshining** [comment]
Okay, as this is it compiles but with a warning and it's not working.

We need to calculate the profit share before any plundering or combat takes place, and remember that when it does.  @Janaszar suggested putting the line to do so here:
https://github.com/kikotheexile/Endless-Sky-Civil-War/pull/170/files#diff-bd5c5f8ac293c62c58b8640fd06c248253fa442079a27045cebe642606662966R65

and I've tried a couple of other different places, with no luck.  On compile it complains that the variable is unused.... which I take it means that it's local to that function and won't be remembered down in `BoardingPanel::KeyDown`?  Even though I put it in the class here:
https://github.com/kikotheexile/Endless-Sky-Civil-War/pull/170/files#diff-144840faa9bcf20eb39fb5de9b8c87370b5dae5cb53fd33972388af58f223f09R130

So it's forgetting it and then getting zero for the profitShareRatio when actually needed which results in zero charges.

Anyway it's 0245 now.  Hopefully this is a simple issue that more experience and/or more sleep can solve :-)

File: source/Crew.cpp
Comment (lukearndt) [inline source/Crew.cpp:300]: Cool. Could you please delete `Crew::ShareProfit` as part of this change, and update any places where it's called to use `Crew::CalculateProfitShare` instead?

That way, we won't have the exact same logic duplicated in two places.

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
				" credits to your crew as profit shares."
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp:415]:
```suggestion
				profitShares += profitShareRatio * victim->Cost() * Depreciation::Full();

```

File: source/Crew.cpp
Comment (lukearndt) [inline source/Crew.cpp]:
```suggestion
double Crew::CalculateProfitShareRatio(const std::vector<std::shared_ptr<Ship>> &ships, const Ship * flagship)
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
	profitShareRatio = Crew::CalculateProfitShareRatio(player.Ships(), player.Flagship());
```

File: source/Crew.h
Comment (lukearndt) [inline source/Crew.h]:
```suggestion
	// Calculate the proportion of profit that must be shared with the fleet
	static double CalculateProfitShareRatio(const std::vector<std::shared_ptr<Ship>> &ships, const Ship * flagship);
```

File: source/Account.h
Comment (lukearndt) [inline source/Account.h:57]:
```suggestion
	void AddProfitShare(int64_t amount);
```

File: source/Account.cpp
Comment (lukearndt) [inline source/Account.cpp:353]:
```suggestion
void Account::AddProfitShare(int64_t amount)
```

File: source/Account.cpp
Comment (lukearndt) [reply source/Account.cpp]:
```suggestion
// Shared profits that the player owes the fleet, but hasn't paid yet.
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [reply source/BoardingPanel.cpp:242]:
```suggestion
			player.Accounts().AddProfitShare(profitShares);
```

File: source/BoardingPanel.cpp
Comment (lukearndt) [inline source/BoardingPanel.cpp]:
```suggestion
				"You must share " +
				Format::Number(profitShares) +
				" credits of this profit with your crew."
```

**Review by lukearndt — APPROVED** [review]

**Comment by ravenshining** [comment]
Everything checks out, I think we can merge this now!

**Comment by kikotheexile** [comment]
I'll leave this to @lukearndt

**Pull request merged**
